File: android/app/src/main/python/stock_monitoring_app/indicators/atr_indicator.py
import pandas as pd
from .base_indicator import Indicator

class ATRIndicator(Indicator):
    """
    Calculates Average True Range (ATR) using only pandas.
    """

    @staticmethod
    def get_search_space():
        return {
            "window": [10, 14, 20],
        }


    def __init__(
        self,
        df: pd.DataFrame,
        window: int = 14
    ):
        """
        Args:            df: OHLCV DataFrame. Must contain 'High', 'Low', 'Close'.
            window: The time period for ATR calculation.
        """
        super().__init__(df)
        self.window = window

        if len(self.df) < self.window + 1: # Needs window + 1 for prev_close and rolling
            raise ValueError(
                f"Insufficient data for {self.__class__.__name__} (window: {self.window}): "
                f"{len(self.df)} rows provided, requires at least {self.window + 1} rows."
            )

        self.atr_col_name = f'ATR_{self.window}'
        self.low_atr_signal_col = f'ATR_Low_Signal_{self.window}'
        self.high_atr_signal_col = f'ATR_High_Signal_{self.window}'

        # Example signal orientations:
        # Buy on low ATR (anticipating volatility increase / end of consolidation)
        # Sell on high ATR (anticipating volatility decrease / exhaustion)
        self.signal_orientations[self.low_atr_signal_col] = 'buy'
        self.signal_orientations[self.high_atr_signal_col] = 'sell'

    def calculate(self) -> pd.DataFrame:
        """
        Calculates ATR and adds 'ATR_<window>' column.
        Also adds example 'ATR_Low_Signal_<window>' and 'ATR_High_Signal_<window>' columns.
        """
        high = self.df['High']
        low = self.df['Low']
        close = self.df['Close']
        prev_close = close.shift(1)

        # Calculate True Range (TR)
        tr1 = high - low
        tr2 = (high - prev_close).abs()
        tr3 = (low - prev_close).abs()
        tr_series = pd.concat([tr1, tr2, tr3], axis=1).max(axis=1)
        
        # Calculate ATR
        atr_series = tr_series.rolling(window=self.window, min_periods=self.window).mean()
        self.df[self.atr_col_name] = atr_series

        # Generate example signals based on ATR relative to its own median
        # This ensures that changing 'window' can change the signals
        if not atr_series.dropna().empty:

            # Using a rolling median for a more dynamic threshold.
            # The window for the rolling median of ATR could be self.window or another value.
            # Using self.window here for consistency with ATR calculation window.
            # Ensure min_periods for median is reasonable, e.g., at least half the window or a fixed number.
            rolling_median_window = max(1, self.window // 2) # Example: half of ATR window, min 1
            atr_median_threshold = atr_series.rolling(window=self.window, min_periods=rolling_median_window).median()
            
            self.df[self.low_atr_signal_col] = (atr_series < atr_median_threshold)
            self.df[self.high_atr_signal_col] = (atr_series > atr_median_threshold)
            
            self.df[self.low_atr_signal_col] = self.df[self.low_atr_signal_col].fillna(False).astype(bool)
            self.df[self.high_atr_signal_col] = self.df[self.high_atr_signal_col].fillna(False).astype(bool)
        else:
            self.df[self.low_atr_signal_col] = False
            self.df[self.high_atr_signal_col] = False
            
        return self.df

File: src/atr_indicator.cpp
#include "atr_indicator.hpp"
#include <algorithm>
#include <cmath>
#include <numeric>
#include <stdexcept>

// --- Helper rolling mean/median implementations ---

std::vector<double> rolling_mean(const std::vector<double>& data, int window) {
    std::vector<double> result(data.size(), std::nan(""));
    if (window <= 0 || data.size() < static_cast<size_t>(window)) return result;
    double sum = 0.0;
    for (size_t i = 0; i < data.size(); ++i) {
        sum += data[i];
        if (i >= static_cast<size_t>(window)) sum -= data[i - window];
        if (i + 1 >= static_cast<size_t>(window))
            result[i] = sum / window;
    }
    return result;
}

std::vector<double> rolling_median(const std::vector<double>& data, int window, int min_periods) {
    std::vector<double> result(data.size(), std::nan(""));
    if (window <= 0 || data.empty()) return result;
    for (size_t i = 0; i < data.size(); ++i) {
        size_t start = (i + 1 >= static_cast<size_t>(window)) ? i + 1 - window : 0;
        size_t len = i + 1 - start;
        if (static_cast<int>(len) < min_periods) continue;
        std::vector<double> window_data(data.begin() + start, data.begin() + i + 1);
        std::vector<double> valid;
        for (auto v : window_data) if (!std::isnan(v)) valid.push_back(v);
        if (valid.empty()) continue;
        std::sort(valid.begin(), valid.end());
        size_t n = valid.size();
        if (n % 2 == 0)
            result[i] = (valid[n / 2 - 1] + valid[n / 2]) / 2.0;
        else
            result[i] = valid[n / 2];
    }
    return result;
}

// --- ATRIndicator implementation ---

ATRIndicator::ATRIndicator(const DataFrame& df, int window)
    : Indicator(df), window_(window)
{
    if (df_.close.size() < static_cast<size_t>(window_) + 1)
        throw std::invalid_argument("Insufficient data for ATRIndicator (window: " + std::to_string(window_) + "): "
            + std::to_string(df_.close.size()) + " rows provided, requires at least " + std::to_string(window_ + 1) + " rows.");

    atr_col_name_ = "ATR_" + std::to_string(window_);
    low_atr_signal_col_ = "ATR_Low_Signal_" + std::to_string(window_);
    high_atr_signal_col_ = "ATR_High_Signal_" + std::to_string(window_);

    signal_orientations_[low_atr_signal_col_] = "buy";
    signal_orientations_[high_atr_signal_col_] = "sell";
}

DataFrame ATRIndicator::calculate() {
    size_t n = df_.close.size();
    std::vector<double>& high = df_.high;
    std::vector<double>& low = df_.low;
    std::vector<double>& close = df_.close;
    std::vector<double> prev_close(n, std::nan(""));
    for (size_t i = 1; i < n; ++i)
        prev_close[i] = close[i - 1];

    // Calculate True Range (TR)
    std::vector<double> tr1(n), tr2(n), tr3(n), tr(n, std::nan(""));
    for (size_t i = 0; i < n; ++i) {
        tr1[i] = high[i] - low[i];
        tr2[i] = (i > 0) ? std::fabs(high[i] - prev_close[i]) : std::nan("");
        tr3[i] = (i > 0) ? std::fabs(low[i] - prev_close[i]) : std::nan("");
        tr[i] = std::max({tr1[i], tr2[i], tr3[i]});
    }

    // ATR as rolling mean of TR
    std::vector<double> atr = rolling_mean(tr, window_);

    // Add ATR column to DataFrame (add as an extra field for this indicator)
    df_.extra_columns[atr_col_name_] = atr;

    // Rolling median for dynamic threshold
    int rolling_median_window = std::max(1, window_ / 2);
    std::vector<double> atr_median = rolling_median(atr, window_, rolling_median_window);

    // Build signals: ATR below/above its rolling median
    std::vector<bool> low_signal(n, false), high_signal(n, false);
    for (size_t i = 0; i < n; ++i) {
        if (!std::isnan(atr[i]) && !std::isnan(atr_median[i])) {
            low_signal[i] = (atr[i] < atr_median[i]);
            high_signal[i] = (atr[i] > atr_median[i]);
        }
    }
    df_.extra_bool_columns[low_atr_signal_col_] = low_signal;
    df_.extra_bool_columns[high_atr_signal_col_] = high_signal;

    return df_;
}
